### PR TITLE
property: default properties don't work with a NULL query.

### LINKED ISSUE #11594: EVP_set_default_properties() does not work
The function `EVP_set_default_properties()` does not seem to be working. I created the following test program:

````
#include <stdio.h>
#include <openssl/evp.h>

int main(void)
{
    EVP_MD *md;

    if (!EVP_set_default_properties(NULL, "provider=wibble"))
        printf("Failed to set default properties\n");

    md = EVP_MD_fetch(NULL, "SHA2-256", NULL);
    if (md == NULL)
        printf("Failed to fetch the digest\n");
    else
        printf("Succeeded fetching the digest\n");

    EVP_MD_free(md);

    return 0;
}

````

The output I received was "Succeeded fetching the digest". But I expected it to fail, because no implementations match the specified criteria.

I also note that `EVP_set_default_properties()` does not exist in the test suite - so we should fix that.

### PR DESCRIPTION
**Pull request by paulidale**
Specifying an algorithm with a NULL query ignores the default properties.

Fixes #11594

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/property/property.c [hunk 1]
```diff
@@ -324,27 +324,31 @@ int ossl_method_store_fetch(OSSL_METHOD_STORE *store, int nid,
...
     if (prop_query == NULL) {
-        if ((impl = sk_IMPLEMENTATION_value(alg->impls, 0)) != NULL) {
-            best_method = &impl->method;
-            ret = 1;
+        if (store->global_properties == NULL) {
+            if ((impl = sk_IMPLEMENTATION_value(alg->impls, 0)) != NULL) {
+                best_method = &impl->method;
+                ret = 1;
+            }
+            goto fin;
```
Comment (levitte) [inline crypto/property/property.c:332]: Having this bit smack in the middle of property query processing makes for a confusing read, at least to me.  May I suggest this rearrangement?

``` C
    if (prop_query != NULL) {
        p2 = pq = ossl_parse_query(store->ctx, prop_query);
    }
    if (store->global_properties != NULL) {
        if (pq == NULL) {
            pq = store->global_properties;
        } else {
            p2 = ossl_property_merge(pq, store->global_properties);
            if (p2 == NULL)
                goto fin;
            ossl_property_free(pq);
            pq = p2;
        }
    }

    if (pq == NULL) {
        if ((impl = sk_IMPLEMENTATION_value(alg->impls, 0)) != NULL) {
            best_method = &impl->method;
            ret = 1;
        }
        goto fin;
    }
    optional = ossl_property_has_optional(pq);
    for (j = 0; j < sk_IMPLEMENTATION_num(alg->impls); j++) {
```
```

File: crypto/property/property.c [hunk 2]
```diff
@@ -358,7 +362,7 @@ int ossl_method_store_fetch(OSSL_METHOD_STORE *store, int nid,
     else
         ret = 0;
     ossl_property_unlock(store);
-    ossl_property_free(pq);
+    ossl_property_free(p2);
```
Comment (levitte) [inline crypto/property/property.c:365]: Make sure to initialise `p2`

Comment (paulidale) [reply crypto/property/property.c:365]: `p2` is initialised to NULL above.

I welcome suggestions and would love a patch commit to correct my mis-comprehensions.

The bottom line is that I understand the property code well enough but I do not understand the higher levels.

Comment (paulidale) [reply crypto/property/property.c:365]: I.e. add a patch here or create a new PR.

Comment (levitte) [reply crypto/property/property.c:365]: D'oh, I'm blind...

File: test/evp_extra_test.c [hunk 3]
```diff
@@ -471,6 +471,35 @@ static EVP_PKEY *load_example_hmac_key(void)
     return pkey;
 }
 
+static int test_EVP_set_default_properties(void)
+{
+    OPENSSL_CTX *ctx;
+    EVP_MD *md;
```
Comment (levitte) [inline test/evp_extra_test.c:477]: Travis tells us that `md` needs to be initialised

**Comment by levitte** [comment]
> I welcome suggestions and would love a patch commit to correct my misscomprehensions.

Done

**Comment by paulidale** [comment]
@levitte I'm beyond being able to reason about the code paths tonight.  Feel free to push changes to this branch and I'll review them tomorrow.

**Comment by levitte** [comment]
> @levitte I'm beyond being able to reason about the code paths tonight. Feel free to push changes to this branch and I'll review them tomorrow.

Will do

File: crypto/property/property.c [hunk 4]
```diff
@@ -323,28 +323,33 @@ int ossl_method_store_fetch(OSSL_METHOD_STORE *store, int nid,
...
-            goto fin;
-        ossl_property_free(pq);
-        pq = p2;
-    }
     optional = ossl_property_has_optional(pq);
     for (j = 0; j < sk_IMPLEMENTATION_num(alg->impls); j++) {
         impl = sk_IMPLEMENTATION_value(alg->impls, j);
         score = ossl_property_match_count(pq, impl->properties);
-        if (score > best) {
+        if (score > best && score >= 0) {
```
Comment (paulidale) [inline crypto/property/property.c:352]: Note to self, the `score >= 0` check is redundant.  Negative scores mean no match found and this equals the starting value for best -- thus the condition is always false.

**Comment by paulidale** [comment]
Thanks for the rework.  Much appreciated.

**Comment by paulidale** [comment]
I'm not entirely sure I like this fix (the final commit), it boils down to the default properties being associated with the store not the library context.

**Comment by levitte** [comment]
I agree. The default props should be stored directly in the library context. I think that was already mentioned in #11594

**Comment by paulidale** [comment]
I'm happy to pull the last commit in favour of a better fix.

**Comment by paulidale** [comment]
The default property query is now per library context instead of per store.

**Comment by paulidale** [comment]
Ping? @levitte ?

**Comment by paulidale** [comment]
Ping @openssl/committers

**Comment by paulidale** [comment]
Merged to master thanks.

**Pull request closed without merging**